File: plugin/hooks/get-posttool-output.py
#!/usr/bin/env python3
"""
get-posttool-output.py - Unified PostToolUse hook for all tools

TRIGGER: PostToolUse (no matcher - runs for all tools)

Consolidates general PostToolUse hooks into a single Python dispatcher.
For Bash-specific PostToolUse hooks, see get-bash-posttool-output.py.

Handlers can:
- Warn about tool results (return {"warning": "..."})
- Inject additional context (return {"additionalContext": "..."})
- Allow silently (return None)
"""

import json
import sys
from pathlib import Path

# Add handlers directory to path
SCRIPT_DIR = Path(__file__).parent
sys.path.insert(0, str(SCRIPT_DIR))

from posttool_handlers import get_all_handlers


def read_hook_input() -> dict:
    """Read and parse JSON input from stdin."""
    if sys.stdin.isatty():
        return {}

    try:
        raw = sys.stdin.read()
        return json.loads(raw) if raw.strip() else {}
    except json.JSONDecodeError:
        return {}


def main():
    # Read hook input
    hook_data = read_hook_input()

    # Extract tool info
    tool_name = hook_data.get("tool_name", "")
    tool_result = hook_data.get("tool_result", {}) or hook_data.get("tool_response", {})
    session_id = hook_data.get("session_id", "")

    if not tool_name:
        print("{}")
        return

    # Build context for handlers
    context = {
        "session_id": session_id,
        "hook_data": hook_data,
    }

    # Run all handlers
    warnings = []
    additional_contexts = []

    for handler in get_all_handlers():
        try:
            result = handler.check(tool_name, tool_result, context)
            if result:
                if result.get("warning"):
                    warnings.append(result["warning"])
                if result.get("additionalContext"):
                    additional_contexts.append(result["additionalContext"])
        except Exception as e:
            sys.stderr.write(f"get-posttool-output: handler error: {e}\n")

    # Output warnings to stderr
    if warnings:
        for warning in warnings:
            sys.stderr.write(warning + "\n")

    # Build response with additionalContext if present
    if additional_contexts:
        response = {
            "hookSpecificOutput": {
                "hookEventName": "PostToolUse",
                "additionalContext": "\n\n".join(additional_contexts)
            }
        }
        print(json.dumps(response))
    else:
        print("{}")


if __name__ == "__main__":
    main()

File: plugin/hooks/src/io/github/cowwoc/cat/hooks/GetPosttoolOutput.java
package io.github.cowwoc.cat.hooks;

import static io.github.cowwoc.requirements13.java.DefaultJavaValidators.requireThat;

import io.github.cowwoc.cat.hooks.tool.post.AutoLearnMistakes;
import tools.jackson.databind.JsonNode;

import java.util.ArrayList;
import java.util.List;

/**
 * get-posttool-output - Unified PostToolUse hook for all tools
 *
 * TRIGGER: PostToolUse (no matcher - runs for all tools)
 *
 * Consolidates general PostToolUse hooks into a single Java dispatcher.
 * For Bash-specific PostToolUse hooks, see GetBashPosttoolOutput.
 *
 * Handlers can:
 * - Warn about tool results (return warning string)
 * - Inject additional context (return additionalContext)
 * - Allow silently (return null)
 */
public final class GetPosttoolOutput
{
  private static final List<PosttoolHandler> HANDLERS = List.of(
      new AutoLearnMistakes());

  private GetPosttoolOutput()
  {
    // Utility class
  }

  /**
   * Entry point for the general posttool output hook.
   *
   * @param _args command line arguments (unused)
   */
  @SuppressWarnings("UnusedVariable")
  public static void main(String[] _args)
  {
    HookInput input = HookInput.readFromStdin();

    String toolName = input.getToolName();
    if (toolName.isEmpty())
    {
      HookOutput.empty();
      return;
    }

    JsonNode toolResult = input.getToolResult();
    String sessionId = input.getSessionId();
    requireThat(sessionId, "sessionId").isNotBlank();

    List<String> warnings = new ArrayList<>();
    List<String> additionalContexts = new ArrayList<>();

    // Run all general posttool handlers
    for (PosttoolHandler handler : HANDLERS)
    {
      try
      {
        PosttoolHandler.Result result = handler.check(toolName, toolResult, sessionId, input.getRaw());
        if (!result.warning().isEmpty())
          warnings.add(result.warning());
        if (!result.additionalContext().isEmpty())
          additionalContexts.add(result.additionalContext());
      }
      catch (Exception e)
      {
        System.err.println("get-posttool-output: handler error: " + e.getMessage());
      }
    }

    // Output warnings to stderr
    for (String warning : warnings)
    {
      System.err.println(warning);
    }

    // Build response with additionalContext if present
    if (!additionalContexts.isEmpty())
    {
      String combined = String.join("\n\n", additionalContexts);
      HookOutput.additionalContext("PostToolUse", combined);
    }
    else
    {
      HookOutput.empty();
    }
  }
}
